### PR TITLE
Code analysis with static analysers based on microsoft rules

### PR DESCRIPTION
**Pull request by HeroMaxPower**
## Description
Provided some code cleanup based on some microsoft rules, 

CA1507 Use nameof in place of string literal
CA1827 Count() is used where Any() could be used instead to improve performance
CA1068 Method should take CancellationToken as the last parameter
CA1834 Use 'StringBuilder.Append(char)' instead of 'StringBuilder.Append(string)' when the input is a constant unit string
CA2208 Replace this argument with one of the method's parameter names. Note that the provided parameter name should have the exact casing as declared on the method.
CA1816 Change .Dispose() to call GC.SuppressFinalize(object). This will prevent derived types that introduce a finalizer from needing to re-implement 'IDisposable' to call it.

S3626 Remove this redundant jump.
S3415 Make sure these 2 arguments are in the correct order: expected value, actual value.
S2971 Use 'Count' property here instead.
S1155 Use '.Any()' to test whether is empty or not.
S1116 Remove this empty statement

## Related issue
Related to #2827

### CODE REVIEW

**Comment by HeroMaxPower** [comment]
@Haplois got some more for you.

I guess the most contentious part is the GC collection rule CA1816 Change .Dispose() to call GC.SuppressFinalize(object). This will prevent derived types that introduce a finalizer from needing to re-implement 'IDisposable' to call it.. I think everything is fine but i could be wrong.

**Comment by HeroMaxPower** [comment]
Hi @Haplois im not sure what failed here, i don't think it's related to my code but package restoring. It might need another run
last error message was:

``scripts/build.sh: line 395:  5912 Segmentation fault      (core dumped) $dotnet pack --no-build $stagingDir/${i} -o $packageOutputDir -p:Version=$TPB_Version`
`

File: src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs [hunk 1]
```diff
@@ -336,7 +336,6 @@ private void PopulateHtmlFile()
                 EqtTrace.Error("HtmlLogger : Failed to populate html file. Exception : {0}",
                     ex.ToString());
                 ConsoleOutput.Instance.Error(false, string.Concat(HtmlResource.HtmlLoggerError), ex.Message);
-                return;
```
Comment (Haplois) [inline src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:339]: Any reason for this removal? It changes the current behavior.

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs [hunk 2]
```diff
@@ -35,9 +35,11 @@ public void Dispose()
 #if NETFRAMEWORK
             // tcpClient.Close() calls tcpClient.Dispose().
             this.tcpClient?.Close();
+            GC.SuppressFinalize(this);
 #else
             // tcpClient.Close() not available for netcoreapp1.0
             this.tcpClient?.Dispose();
+            GC.SuppressFinalize(this);
```
Comment (Haplois) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs:42]: nit: We can remove this call and add it outside of `#if` preprocessor directive, just after `#endif`.

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs [hunk 3]
```diff
@@ -35,9 +35,11 @@ public void Dispose()
 #if NETFRAMEWORK
             // tcpClient.Close() calls tcpClient.Dispose().
             this.tcpClient?.Close();
+            GC.SuppressFinalize(this);
```
Comment (Haplois) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs:38]: see the next comment.

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketCommunicationManagerTests.cs [hunk 4]
```diff
@@ -45,12 +45,14 @@ public void Dispose()
 #if NETFRAMEWORK
             // tcpClient.Close() calls tcpClient.Dispose().
             this.tcpClient?.Close();
+            GC.SuppressFinalize(this);
```
Comment (Haplois) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketCommunicationManagerTests.cs:48]:
```suggestion
```

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs [hunk 5]
```diff
@@ -38,9 +38,11 @@ public void Dispose()
 #if NETFRAMEWORK
             // tcpClient.Close() calls tcpClient.Dispose().
             this.tcpClient?.Close();
+            GC.SuppressFinalize(this);
```
Comment (Haplois) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs:41]:
```suggestion
```

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs [hunk 6]
```diff
@@ -38,9 +38,11 @@ public void Dispose()
...
 #else
             // tcpClient.Close() not available for netcoreapp1.0
             this.tcpClient?.Dispose();
+            GC.SuppressFinalize(this);
```
Comment (Haplois) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs:45]: nit: Move this after `#endif`.

**Review by Haplois — CHANGES_REQUESTED** [review]
Thank you, @HeroMaxPower, for your contributions. LGTM, with some nitpicking and questions.

File: src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs (hunk 1 above)
Comment (HeroMaxPower) [reply src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs]: yes you are right. reverting.

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs (hunk 2 above)
Comment (HeroMaxPower) [reply test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs]: resolved added GC.SuppressFinalize(this); after the #endif

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketCommunicationManagerTests.cs (hunk 4 above)
Comment (HeroMaxPower) [reply test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketCommunicationManagerTests.cs]: resolved added GC.SuppressFinalize(this); after the #endif

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs (hunk 3 above)
Comment (HeroMaxPower) [reply test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs]: resolved added GC.SuppressFinalize(this); after the #endif

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs (hunk 5 above)
Comment (HeroMaxPower) [reply test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs]: resolved added GC.SuppressFinalize(this); after the #endif

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs (hunk 6 above)
Comment (HeroMaxPower) [reply test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketClientTests.cs]: resolved added GC.SuppressFinalize(this); after the #endif

**Review by Haplois — APPROVED** [review]
LGTM.

**Comment by Haplois** [comment]
@HeroMaxPower, this is merged. Thank you for all your contributions lately.

**Pull request merged**